pilot/pkg/networking/plugin/authz/rbac_test.go

```diff
@@ -529,3 +529,187 @@ func TestConvertRbacRulesToFilterConfig(t *testing.T) {
+			Spec:       generateServiceRole([]string{"service"}, []string{"POST"}),
+		},
+	}
+	bindings := []model.Config{
+		{
+			ConfigMeta: model.ConfigMeta{Name: "service-role-binding-1"},
+			Spec:       generateServiceBinding("user1", "service-role-1", rbacproto.EnforcementMode_ENFORCED),
+		},
+		{
+			ConfigMeta: model.ConfigMeta{Name: "service-role-binding-2"},
+			Spec:       generateServiceBinding("user2", "service-role-1", rbacproto.EnforcementMode_PERMISSIVE),
+		},
+		{
+			ConfigMeta: model.ConfigMeta{Name: "service-role-binding-3"},
+			Spec:       generateServiceBinding("user3", "service-role-2", rbacproto.EnforcementMode_ENFORCED),
+		},
+		{
+			ConfigMeta: model.ConfigMeta{Name: "service-role-binding-4"},
+			Spec:       generateServiceBinding("user4", "service-role-2", rbacproto.EnforcementMode_PERMISSIVE),
+		},
+	}
+
+	enforcePolicy1 := &policy.Policy{
+		Permissions: []*policy.Permission{generatePermission(":method", "GET")},
+		Principals:  []*policy.Principal{generatePrincipal("user1")},
+	}
+
+	enforcePolicy2 := &policy.Policy{
+		Permissions: []*policy.Permission{generatePermission(":method", "POST")},
+		Principals:  []*policy.Principal{generatePrincipal("user3")},
+	}
+
+	permissivePolicy1 := &policy.Policy{
+		Permissions: []*policy.Permission{generatePermission(":method", "GET")},
+		Principals:  []*policy.Principal{generatePrincipal("user2")},
+	}
+
+	permissivePolicy2 := &policy.Policy{
+		Permissions: []*policy.Permission{generatePermission(":method", "POST")},
+		Principals:  []*policy.Principal{generatePrincipal("user4")},
+	}
+
+	enforcedRbac := &policy.RBAC{
+		Action: policy.RBAC_ALLOW,
+		Policies: map[string]*policy.Policy{
+			"service-role-1": enforcePolicy1,
+			"service-role-2": enforcePolicy2,
+		},
+	}
+	shadowRbac := &policy.RBAC{
+		Action: policy.RBAC_ALLOW,
+		Policies: map[string]*policy.Policy{
+			"service-role-1": permissivePolicy1,
+			"service-role-2": permissivePolicy2,
+		},
+	}
+	testCases := []struct {
+		name       string
+		service    string
+		roles      []model.Config
+		bindings   []model.Config
+		rbac       *policy.RBAC
+		shadowRbac *policy.RBAC
+	}{
+		{
+			name:       "exact matched service",
+			service:    "service",
+			roles:      roles,
+			bindings:   bindings,
+			rbac:       enforcedRbac,
+			shadowRbac: shadowRbac,
+		},
+		{
+			name:     "empty roles",
+			service:  "service",
+			roles:    []model.Config{},
+			bindings: bindings,
+		},
+		{
+			name:     "empty bindings",
+			service:  "service",
+			roles:    roles,
+			bindings: []model.Config{},
+		},
+	}
+
+	for _, tc := range testCases {
+		rbac := convertRbacRulesToFilterConfig(tc.service, tc.roles, tc.bindings)
+		if len(tc.roles) == 0 || len(tc.bindings) == 0 {
+			if got, want := len(rbac.Rules.Policies), 0; got != want {
+				t.Errorf("len(rbac.Rules.Policies), got: %d, want: %d", got, want)
+			}
+		} else {
+			if !reflect.DeepEqual(*tc.rbac, *rbac.Rules) {
```

Comment:
Is there any part of rbac you don't want to compare? In other words, why don't use one RBAC golden data that have both rules and shadow?